### PR TITLE
Z

### PR DESCRIPTION
**Pull request by jasonmotha**

### CODE REVIEW

File: src/constants.h [hunk 1]
```diff
@@ -24,7 +24,7 @@
 #define MAX_NUM_BASIN_SURFACES 12 // the maximum number of basin surfaces
 #define MAX_NUM_1D_VELOCITY_MODELS 1 // the maximum number of 1D velocity models
 #define MAX_NUM_BASIN_BOUNDARIES 3 // the maximum number of boundaries per velocity model
-#define MAX_NUM_GEN_MULTI_PROFILES 1000 // the maximum number of profiles able to be generated by GEN_MULTI_PROFILES_CALL
+#define MAX_NUM_GEN_MULTI_PROFILES 27000 // the maximum number of profiles able to be generated by GEN_MULTI_PROFILES_CALL
```
Comment (jpa198) [inline src/constants.h:27]: The number of virtual stations we use?

File: Velocity_Model/z.py [hunk 2]
```diff
@@ -0,0 +1,173 @@
+"""
+Library functions for extracting Z values from the NZVM in a pythonic manner
+Supports NZVM versions 2.03, 2.04, 2.06
+"""
+import os
```
Comment (jpa198) [inline Velocity_Model/z.py:5]: You can narrow this import to only 'path'

File: Velocity_Model/z.py [hunk 3]
```diff
@@ -0,0 +1,173 @@
...
+
+IN_BASIN_SIGMA = 0.3
+
+file_output_format = {"Z1.0": "Z/Z_1.0.txt", "Z2.5": "Z/Z_2.5.txt"}
+column_names = {"Z1.0": "Z_1.0(km)", "Z2.5": "Z_2.5(km)"}
+
+
+def get_basin_full_path(basin_list):
+    vm_dir = os.path.abspath(
+        os.path.join(os.path.dirname(os.path.realpath(__file__)), "../")
```
Comment (jpa198) [inline Velocity_Model/z.py:25]: Final / doesn't need to be there? And having an explicit separator defeats the purpose of using join?

File: Velocity_Model/z.py [hunk 4]
```diff
@@ -0,0 +1,173 @@
...
+        "Data/USER20_BASINS/CollingwoodBasinOutline_2_WGS84_v1.txt",
+        "Data/USER20_BASINS/CollingwoodBasinOutline_3_WGS84_v1.txt",
+    ]
+)
+basin_outlines = {
+    "2.03": v203_basin_list,
+    "2.04": v204_basin_list,
+    "2.05": v205_basin_list,
+    "2.06": v206_basin_list,
+}
```
Comment (jpa198) [inline Velocity_Model/z.py:78]: Maybe add a version for 1.66? (Or what ever the last 1.XX version is)
I like how you've used the incremental nature of the basins though.

File: Velocity_Model/z.py [hunk 5]
```diff
@@ -0,0 +1,173 @@
...
+    """
+    Returns a np array of len(stat_df) containing 0.3 or 0.5 depending if the station is within a basin or not
+
+    Loops over the respective basin files for each NZVM version
+    """
+    sigma_df = stat_df.copy()
+    sigma_df["in_basin_flag"] = False
+    sigma_df = sigma_df[
+        ["in_basin_flag"]
+    ]  # drops all columns other than the flag and the index
```
Comment (jpa198) [inline Velocity_Model/z.py:164]: sigma_df = pd.DataFrame(index=stat_df.index)
sigma_df["in_basin_flag"] = False

Would be more efficient than copying the whole dataframe then dropping most of it? Or at least less lines?

File: Velocity_Model/scripts/get_z.py [hunk 6]
```diff
@@ -0,0 +1,31 @@
...
+        "--z-type",
+        choices=[
+            "Z1.0",
+            "Z2.5",
+        ],  # , "vs30", "vs500"], -- couldn't get vs30 or vs500 working despite documentation saying it should work
+        help="Selects what Z type to extract: [%(choices)s]",
+        nargs="+",
+        required=True,
+    )
+    parser.add_argument("-v", "--version", help="VM version to use", default="2.06")
```
Comment (jpa198) [inline Velocity_Model/scripts/get_z.py:22]: This should have a choices setting, maybe from the basin-list dictionary keys? (Velocity_Model.z.basin_outlines)

File: Velocity_Model/scripts/get_z.py [hunk 7]
```diff
@@ -0,0 +1,31 @@
...
+        "-z",
+        "--z-type",
+        choices=[
+            "Z1.0",
+            "Z2.5",
+        ],  # , "vs30", "vs500"], -- couldn't get vs30 or vs500 working despite documentation saying it should work
+        help="Selects what Z type to extract: [%(choices)s]",
+        nargs="+",
+        required=True,
+    )
```
Comment (jpa198) [inline Velocity_Model/scripts/get_z.py:21]: Why are these required flagged arguments, rather than standard positional arguments?

**Review by jpa198 — CHANGES_REQUESTED** [review]
Mostly looks alright, most comments are questions but still a couple of changes

File: Velocity_Model/scripts/station_containment.py [hunk 8]
```diff
@@ -28,5 +28,8 @@
 
     for stat in stat_file.iterrows():
         if path.contains_point((stat[1].lon, stat[1].lat)):
-            #print(f"station: {stat[0], stat[1].lon, stat[1].lat} is contained in the basin {os.path.basename(outline_fp)}")
-            print(stat[1].lon, stat[1].lat, stat[0])
\ No newline at end of file
+            print(
+                stat[1].lon, stat[1].lat, stat[0], True, {os.path.basename(outline_fp)}
```
[the remark this answers is not in the document]
Comment (jasonmotha) [reply Velocity_Model/scripts/station_containment.py:32]: Because I copied and pasted it from an f-string :laughing:

File: src/constants.h (hunk 1 above)
Comment (jasonmotha) [reply src/constants.h:27]: An arbitrarily large number.

File: Velocity_Model/z.py (hunk 4 above)
Comment (jasonmotha) [reply Velocity_Model/z.py:78]: I thought about it, but we don't really use 1.66 anymore - if we did need it for 1.66 we can add it ad-hoc. I think basin stuff is mostly for NZVM 2+ .

It gets a little messier when we incrementally add newer versions of existing basins but not hugely messy.

File: Velocity_Model/scripts/get_z.py (hunk 6 above)
Comment (jasonmotha) [reply Velocity_Model/scripts/get_z.py:22]: Good idea.

File: Velocity_Model/scripts/get_z.py (hunk 7 above)
Comment (jasonmotha) [reply Velocity_Model/scripts/get_z.py:21]: I prefer flagged arguments over positional so it's easy to remember the order / it doesn't matter.

File: Velocity_Model/z.py (hunk 4 above)
Comment (jpa198) [reply Velocity_Model/z.py:78]: Yeah, updating basins was the main issue I had with the current style.
I have an idea about switching over to using dictionaries and replacing keys if we need to do that though.

File: Velocity_Model/z.py [hunk 9]
```diff
@@ -0,0 +1,182 @@
...
+    """
+    Returns a np array of len(stat_df) containing 0.3 or 0.5 depending if the station is within a basin or not
+
+    Loops over the respective basin files for each NZVM version
+    """
+    sigma_df = pd.DataFrame(index=stat_df.index)
+    sigma_df["in_basin_flag"] = False
+    sigma_df = sigma_df[
+        ["in_basin_flag"]
+    ]  # drops all columns other than the flag and the index
```
Comment (jpa198) [inline Velocity_Model/z.py:173]: Deprecated

File: Velocity_Model/z.py [hunk 10]
```diff
@@ -0,0 +1,182 @@
...
+    :param z_types:
+    :param stat_df: Dataframe containing stations - station should be the index and lon and lat should be the columns
+    :param nzvm_path:
+    :param version: NZVM version to extract values for
+    :return: Dataframe containing all z_types, station names and sigma
+    """
+    if len(stat_df) > MAX_NUM_GEN_MULTI_PROFILES:
+        print(
+            "Increase the MAX_NUM_GEN_MULTI_PROFILES constant in the VM code"
+            " (and the constant used in this check) and rerun"
```
Comment (jpa198) [inline Velocity_Model/z.py:101]: State the current value, and also say to recompile the NZVM (I assume it's a hardcoded value rather than in a config file)

**Review by jpa198 — CHANGES_REQUESTED** [review]

File: Velocity_Model/scripts/station_containment.py [hunk 11]
```diff
@@ -28,5 +28,6 @@
 
     for stat in stat_file.iterrows():
         if path.contains_point((stat[1].lon, stat[1].lat)):
-            #print(f"station: {stat[0], stat[1].lon, stat[1].lat} is contained in the basin {os.path.basename(outline_fp)}")
-            print(stat[1].lon, stat[1].lat, stat[0])
\ No newline at end of file
+            print(stat[1].lon, stat[1].lat, stat[0], True, os.path.basename(outline_fp))
```
[the remark this answers is not in the document]
Comment (jasonmotha) [reply Velocity_Model/scripts/station_containment.py:31]: Laziness. :laughing: 

You can just output redirect to save it to a csv. Considering this a quick script adding pd.Series to it probably doubles the complexity.

File: Velocity_Model/z.py (hunk 9 above)
Comment (jpa198) [reply Velocity_Model/z.py:173]: Pls delete this assignment. It launders an already clean dataframe

**Review by claudio525 — APPROVED** [review]

File: Velocity_Model/z.py (hunk 2 above)
Comment (jpa198) [reply Velocity_Model/z.py:5]: This import can be removed

**Review by jpa198 — APPROVED** [review]

File: Velocity_Model/z.py [hunk 12]
```diff
@@ -0,0 +1,178 @@
...
+        "Data/USER20_BASINS/NE_otago/NE_otago_B_outline.txt",
+        "Data/USER20_BASINS/NE_otago/NE_otago_C_outline.txt",
+        "Data/USER20_BASINS/NE_otago/NE_otago_D_outline.txt",
+        "Data/USER20_BASINS/NE_otago/NE_otago_E_outline.txt",
+        "Data/USER20_BASINS/bal_outline_WGS84.txt",
+        "Data/USER20_BASINS/dun_outline_WGS84.txt",
+        "Data/USER20_BASINS/mos_outline_WGS84.txt",
+        "Data/USER20_BASINS/Murchison_Basin_Outline_v1_WGS84.txt",
+        "Data/USER20_BASINS/hakataramea_outline_WGS84.txt",
+        "Data/USER20_BASINS/waitaki_outline_WGS84.txt",
```
Comment (shot0829) [inline Velocity_Model/z.py:59]: just a note, that these lines looks like the exact copy from the c codes.
And at some point in the future we might need to make the NZVM read a config file so other code/scripts can reuse it.

**Review by shot0829 — APPROVED** [review]

File: Velocity_Model/z.py (hunk 12 above)
Comment (jasonmotha) [reply Velocity_Model/z.py:59]: Dibs not. :rofl: :sweat: 

But yeah, sensible suggestion.
